pkg/controller/servicediscovery/privatednsnamespace/hooks.go

```diff
@@ -0,0 +1,160 @@
+package privatednsnamespace
+
+import (
+	"context"
+
+	svcsdk "github.com/aws/aws-sdk-go/service/servicediscovery"
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	cpresource "github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/servicediscovery/v1alpha1"
+	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+func (e *external) observeImpl(ctx context.Context, cpm cpresource.Managed) (managed.ExternalObservation, error) { // nolint:gocyclo
```

Comment:
You should use a seperate `hook` type where you define your custom functions and pass them to `external` which is auto-generated and should not be modified.

Then you can rename this to `observe` as well.

Reply:
updated code to have separate hook struct